### ISSUE TITLE
matplotlib.pyplot.imsave colormaps some grayscale images before saving them

### ISSUE THREAD

**Issue by jni** [matplotlib/matplotlib#3657]
Some people may be using matplotlib as an image IO library. Those people may also be scientists that care about their data being accurately saved on disk. They will be disappointed, however, to find that `imsave` does nothing of the sort:

``` python
In [8]: mpl.pyplot.imsave('io4.png', np.array([[0, 1], [0, 4196]], dtype=np.uint16))

In [10]: im4 = mpl.pyplot.imread('io4.png')

In [11]: im4.shape
Out[11]: (2, 2, 4)

In [12]: np.rollaxis(im4, -1) # display the three channels + alpha of the array
Out[12]: 
array([[[ 0.        ,  0.        ],
        [ 0.        ,  0.49803922]],

       [[ 0.        ,  0.        ],
        [ 0.        ,  0.        ]],

       [[ 0.49803922,  0.49803922],
        [ 0.49803922,  0.        ]],

       [[ 1.        ,  1.        ],
        [ 1.        ,  1.        ]]], dtype=float32)
```

What is happening here is that `imsave` has encountered a data type it's not happy with (uint16, a commonly used data format in microscopy and probably other scientific imaging), and decided to silently colormap the values and save them as a uint8 RGBA image. (See #3616 for a longer discussion.) Adding insult to injury, the color mapping is to jet.

In summary:
- Observed behaviour: imsave colormaps certain images before saving as uint8 RBGA.
- Desired behaviour: either save the data exactly as given, or raise a ValueError, or at the very least raise a warning.

**Comment by Fadel87** [comment]
thanx my friend ... since two days Im trying to solve this myth :D .. thanx a lot :)

**Comment by sanjay-1** [comment]
I've had recent real-world experiences with **matplotlib.pyplot.imsave()** and the issues discussed above.

First, I'm relatively new to Python programming (just a few weeks), though I've been programming in other languages (C++, Java, C, others including many Assembly languages years ago) -- so I'm experienced with programming and APIs.  I was trying to save multiple numpy arrays (each containing either a 2D color or 2D grayscale image).  Each array was read from a file using (what I thought was its complement function), **imread()**.

I encountered the exact problem described here, whereby my saved image was being saved not as grayscale but with some color-mapping.  Since I'm new to Python and one never assumes a professional API/library has major bugs (or inconsistencies), I thus assumed my code and my newness to Python was causing this color-mapping.  I investigated for over an hour, and eventually came to the conclusion my code was not causing this.  More exploring, searching/reading... eventually I asked: "_how can I be sure **imsave()** is doing what I want, maybe there's a missing parameter_".

So I researched further for info on imsave(), and after another hour stumbled onto this issue thread.  I understand the point made by both @jni  and  @efiring.  However, while being inexperienced with pyplot (and Python) though not a complete newbie, I feel imsave() and imread() should be reverse complements of each other (my problem arose because I assumed they were).  This is a reasonable assumption for programmers to make.  Most programmers, especially experienced programmers who are used to working with professional / curated libraries, will assume so.

The ideal solution, one that causes minimal "breakage" to the existing API, would be to modify the imsave() function to first detect whether the image/matrix is 2D grayscale (i.e. has no 3rd tuple), and in that case only to internally force/default the appropriate "cmap" parameter and/optionally "dtype" parameter.  This would make minimal if any change to its API definition, and so should affect no/few existing code.

This problem cost me significant time (over 2 hours already).  Worst of all, it makes me realize that Python library(ies?) I will import in future code may similarly be not necessarily consistent/well-designed.  This I feel is the worst aspect -- for me and possibly for most developers -- to be left with the belief/realization that a specific library/package is "dangerous" or "tricky" to use.  It's counter to why we create libraries/packages.  I not mean to disparage or cast doubts on the people or development community who maintain this/other libraries -- but these are side-effects of such problems existing.

I do not presume to know if there are other issues/constraints around this issue (of whether to change imsave() or not, so I offer this only as one firm data point of one user's recent experiences with this exact issue.

Personally I would prefer to see imread() and imsave() to together act as a no-op, for a given image-type.  At a minimum, the doc pages for both functions should be updated to elucidate the problem.
_______________________________________

If I may make a _related suggestion (for the related doc pages)..._ the "cmap" parameter is not explained well on the [http://matplotlib.org/api/image_api.html#matplotlib.image.imread](matplotlib.org's imread() / imsave()) doc page.  It should better explain the possible (or at least common) values for "cmap".  It says this param defaults to the "_rc image.cmap value_", however even this value is not explained.  I looked for this info in many places, starting with colors.Colormap, but found it explained nowhere.  Only a runtime error msg from the Python interpreter gave me info I needed to proceed.

**Comment by efiring** [comment]
We would be happy to see two PRs, one simple one to improve the documentation as you suggest immediately above, and a second to add the ability to write grayscale pngs as directed by a new kwarg.

**Comment by jni** [comment]
@sanjay-1 my recommendation would be to use scikit-image or imageio for IO. Now that `--pylab` is essentially deprecated, and scikit-image uses PIL by default, there's not much reason to use mpl as an image IO library. Having said that, if you want to put together a PR as suggested by @efiring, I'm sure future generations would be grateful!

**Comment by WeatherGod** [comment]
To address the comment about this experience leaving a bad taste in your
mouth, so to speak, I totally understand. Matplotlib is one of the older
packages in the SciPy community, and has a very strict deprecation &
non-breakable policy. While this is great for scientists, this does mean we
carry a lot of historical baggage. Put on top of that the API originally
came from MatLab, and this is a recipe for confusion among users who are
not used to matlab's idiosyncracies.

Other packages in the SciPy community do have a lot of thought put into
them, and doesn't suffer from as much "organic growth" that mpl has. Not
that they don't have gotchas, I won't lie to you. But I think mpl is
probably the worst in this regard.

**Comment by jklymak** [comment]
Just pinging the dev team @matplotlib - what is the point of imsave at this point?  I think a documentation update would help folks realize its not a straight image dump as you'd get from PIL and close this issue.  

So far as I can tell, its the same as imshow, but done directly on the data buffer, and not placed in axes etc.

**Comment by tacaswell** [comment]
Took 8 years, but we got there @jni !

**Comment by jni** [comment]
😂 😲 👏 👏 👏 

also 👋! 😃

**Comment by ebalogun01** [comment]
Encountered issue with imsave adding a fourth channel to my images and had to read through this thread. This behavior is rather frustrating.

**Comment by rcomer** [comment]
@ebalogun01 if you are experiencing problems with the latest version of Matplotlib, please open a new issue with details.
